test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionTest.scala

```diff
@@ -226,13 +226,100 @@ trait CollectionTest {
     assertEquals(5, coll.size())
     assertIteratorSameElementsAsSet(-45, 0, 12, 32, 42)(coll.iterator())
   }
+
+  @Test def toStringShouldConvertEmptyCollection(): Unit = {
+    val coll = factory.empty[Double]
+
+    val expected = "[]"
+    val result = coll.toString()
+    assertEquals(expected, result)
+  }
+
+  @Test def toStringShouldConvertOneElementCollection(): Unit = {
+    val coll = factory.fromElements[Double](1.01)
+
+    // Javascript displays n.0 as n, so one trailing digit must be non-zero.
+    val expected = "[1.01]"
+    val result = coll.toString()
+    assertEquals(expected, result)
+  }
+
+  @Test def toStringShouldUseCommaSpace(): Unit = {
+
+     // Choose Doubles which display the same in Java and Scala-js.
+    // Javascript displays n.0 as n, so one trailing digit must be non-zero.
+    val elements = Seq(88.42, -23.36, 60.173)
+
+    val coll = factory.fromElements[Double](elements:_*)
+
+    // The order of elements returned by each collection is defined
+    // by the collection. Be prepared to handle the general case of any
+    // order here. Specific collections should test the order they specify.
+    val answers = elements.permutations.map(_.mkString("[", ", ", "]"))
+
+    val expected = answers.toSet
+
+    val result = coll.toString()
+
+    val resultIsValid =  expected.contains(result)
+
+    val msg = "result '" + result +
+        "' not in expected set '" + expected + "'"
+
+    assertTrue(msg, resultIsValid)
+  }
+
+  @Test def toStringShouldHandleNullElements(): Unit = {
+    if (factory.allowsNullElement) {
+       val elements = Seq(-1, -2, null, -3)
+
+       val coll = factory.fromElements[Any](elements:_*)
+
+       val answers = elements.permutations.map(_.mkString("[", ", ", "]"))
+       val expected = answers.toSet
+
+       val result = coll.toString()
+       val resultIsValid =  expected.contains(result)
+
+       val msg = "result '" + result +
+           "' not in expected set '" + expected + "'"
+
+    assertTrue(msg, resultIsValid)
+    }
+  }
+
+  @Test def toStringInCustomClassShouldWork(): Unit = {
+
+    case class Custom(name: String, id: Int) extends Ordered[Custom] {
+      def compare(that: Custom) = this.id - that.id
+    }
+
+    val elements = Seq(Custom("A", 1), Custom("b", 2), Custom("C", 3))
+
+    val coll = factory.fromElements[Custom](elements:_*)
+
+    val result = coll.toString()
+
+    val answers = elements.permutations.map(_.mkString("[", ", ", "]"))
+
+    val expected = answers.toSet
+
+    val resultIsValid =  expected.contains(result)
+
+    val msg = "result '" + result +
+        "' not in expected set '" + expected + "'"
+
+    assertTrue(msg, resultIsValid)
+  }
+
 }
 
 trait CollectionFactory {
   def implementationName: String
   def empty[E: ClassTag]: ju.Collection[E]
   def allowsMutationThroughIterator: Boolean = true
   def allowsNullElementQuery: Boolean = true
+  def allowsNullElement: Boolean = true
```

Comment:
:+1:

Reply:
Someday CollectionFactory is going to need a allowsDuplicateElements() (or better name)
to specify that "1, 2, 1. 1" is allowed.  I ran afoul of this lack when rewriting one
of the tests for allowsNullElement.

allowsDuplicateElements() is outside the scope of this PR, but I thought I would 
mention it for consideration.  If I run afoul again, I may implement & use it.